src/Compatibility/Microsoft.DotNet.ApiCompatibility/AssemblySymbolLoader.cs

```diff
@@ -0,0 +1,299 @@
+            }
+
+            _cSharpCompilation = _cSharpCompilation.AddSyntaxTrees(syntaxTrees);
+
+            LoadFromPaths(referencePaths);
+            return _cSharpCompilation.Assembly;
+        }
+
+        public IEnumerable<IAssemblySymbol> LoadMatchingAssemblies(IEnumerable<IAssemblySymbol> fromAssemblies, IEnumerable<string> searchPaths, bool validateMatchingIdentity = true, bool errorOnMissingAssemblies = true)
+        {
+            if (fromAssemblies == null)
+            {
+                throw new ArgumentNullException(nameof(fromAssemblies));
+            }
+
+            if (searchPaths == null)
+            {
+                throw new ArgumentNullException(nameof(searchPaths));
+            }
+
+            List<IAssemblySymbol> matchingAssemblies = new List<IAssemblySymbol>();
+            foreach (IAssemblySymbol assembly in fromAssemblies)
+            {
+                bool found = false;
+                string name = $"{assembly.Name}.dll";
+                foreach (string directory in searchPaths)
+                {
+                    if (!Directory.Exists(directory))
+                    {
+                        throw new FileNotFoundException($"Matching assembly search directory '{directory}' does not exist", nameof(searchPaths));
+                    }
+
+                    string possiblePath = Path.Combine(directory, name);
+                    if (File.Exists(possiblePath))
+                    {
+                        MetadataReference reference = CreateOrGetMetadataReferenceFromPath(possiblePath);
+                        ISymbol symbol = _cSharpCompilation.GetAssemblyOrModuleSymbol(reference);
+                        if (symbol is IAssemblySymbol matchingAssembly)
+                        {
+                            if (validateMatchingIdentity && !matchingAssembly.Identity.Equals(assembly.Identity))
+                            {
+                                _cSharpCompilation = _cSharpCompilation.RemoveReferences(new[] { reference });
+                                _loadedAssemblies.Remove(name);
+                                continue;
+                            }
+
+                            matchingAssemblies.Add(matchingAssembly);
+                            found = true;
+                            break;
+                        }
+                    }
+                }
+
+                if (errorOnMissingAssemblies && !found)
+                {
+                    string assemblyInfo = validateMatchingIdentity ? assembly.Identity.GetDisplayName() : assembly.Name;
+                    _warnings.Add($"Could not find matching assembly: '{assemblyInfo}' in any of the search directories.");
+                }
+            }
+
+            return matchingAssemblies;
+        }
+
+        private IEnumerable<MetadataReference> LoadFromPaths(IEnumerable<string> paths)
+        {
+            List<MetadataReference> result = new List<MetadataReference>();
```

Comment:
Why allocate List<T> when API returns IEnumerable.  Can you just yield?

Reply:
Good catch. I had it like that and changed it to make debugging easier. Will change back to yield.